Abstract and provide implementations using System domain Client and Runtime to extract state roots from XDM

vedhavyas opened the pull request:
This PR removes the usage of system domain client directly, which we are using to extract state roots from XDM, by providing necessary abstraction to extract state roots using System domain Client and System domain runtime where necessary.

We need this runtime  messenger api to get to the function argument which is `Cross domain message`.  In an ideal scenario, we can just deconstruct the Extrinsic without runtime but that is not safe with refactoring.

We introduce `StateRootExtractor` and also provides implementations using `System Domain Client` with full state and `System Domain Runtime` without any state.  We use the implementation with system domain client wherever possible. In the case where it is not, such as fraud proof verification on primary chain where not all nodes might not system domain client, we expect to use the implementation with runtime as this would require loading the runtime and then running the function. This is already a sunk cost for us since we need to verify fraud proof anyway and we can reuse the executor for loading the runtime.

Discussion: https://github.com/subspace/subspace/issues/1230#issuecomment-1472208912

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

nazar-pc commented on crates/subspace-wasm-tools/src/lib.rs:
```diff
@@ -103,3 +109,57 @@ pub fn read_core_domain_runtime_blob(
     Ok(embedded_runtime_blob.to_vec())
 }
+
+struct RuntimeCodeFetcher<'a> {
```
I'm not sure this is a good place for such code. This is crate with tools for building WASM modules (see crate description). While added logic is basically runtime handling logic.

vedhavyas replied:
not sure I agree with that though I do i agree that description needs to change. The added functionality is fairly generic that IMO should belong to WASM tools or utils crate.

nazar-pc commented on domains/client/domain-executor/src/xdm_verifier.rs:
```diff
@@ -99,18 +108,24 @@ where
+        SExecutorDispatch,
+    >(
+        system_domain_wasm.as_ref(),
+        "MessengerApi_extract_xdm_proof_state_roots",
```
Not a fan on such string arguments, but I don't necessarily have a better alternative. The concern is that these functions are easy to break during refactoring and such with no compiler help to detect that.

vedhavyas replied:
I agree on not able to ensure compatibility but I dont see any alternative without bringing in unneeded logic to primary runtime to achieve something similar. I'm still thinking on how to ensure this wont break on refactor. No strong solutions yet.

As for string arguments, TBH, I kinda like it. Though you can make it nicer with some macros like substrate does.

nazar-pc commented on domains/client/domain-executor/src/xdm_verifier.rs:
```diff
@@ -85,32 +92,35 @@ where
+    let system_domain_wasm = primary_api.system_domain_wasm_bundle(primary_best_hash)?;
+
+    let messenger_api =
+        MessengerApiCompatStateless::<SExecutorDispatch>::with_runtime(system_domain_wasm.to_vec());
```
How often is this called? Feels potentially expensive to allocate runtime blob (non-trivial amount of data), maybe worth caching it somewhere rather than instantiating every single time? Substrate internally does caching when you use runtime APIs, but here we're on our own.

nazar-pc reviewed (commented):
BTW, why do we need to either instantiate runtime or have access to the client. If the goal is to decouple things (and we didn't remove any dependencies here) then maybe we can define a trait that something external needs to implement instead? That way we can decouple things from each other, use something trivial in tests and wire everything together at the node level instead.

liuchengxu reviewed (commented):
The changes make sense overall. One thing I noticed is that https://github.com/subspace/subspace/blob/949163a0715ccce821bbbedbe897ef9d539a190c/domains/pallets/messenger/src/lib.rs#L821 could just be a method of `CrossDomainMessage` so that it can be more obvious that this runtime api is stateless and we merely rely on the runtime to extract the xdm from the extrinsic.

nazar-pc commented on domains/client/domain-executor/src/xdm_verifier.rs:
```diff
@@ -283,3 +292,98 @@ where
+
+impl<Executor> FetchRuntimeCode for MessengerApiCompatStateless<Executor> {
+    fn fetch_runtime_code(&self) -> Option<Cow<[u8]>> {
+        Some(self.runtime.as_slice().into())
```
Why not storing `Cow<'static, [u8]>` then?

nazar-pc commented on domains/client/domain-executor/src/xdm_verifier.rs:
```diff
@@ -85,32 +90,38 @@ where
+    let messenger_api = MessengerApiCompatStateless::<Executor>::with_runtime(
+        system_domain_wasm.to_vec(),
+        executor,
+    );
```
I just checked Substrate sources and `self.executor.call()` should cache instances based on hash, but here you have hardcoded static string for hash. If such string is used in other places, we may get two different runtimes with the same hash that will cause hard to debug issues. I think it it worth hashing beforehand.

I really don't like that we're doing heap allocation here with large runtime every time and require hashing of it afterwards too. This should be cached somewhere and reused as much as possible.

liuchengxu replied:
Agree on the frequent large heap allocation as it will be invoked for every xdm extrinsic on the system domain.

vedhavyas replied:
The heap allocation could completely avoided except the first time since we are using the executor that automatically cache versioned runtime and if we have a static hash that is reused for this purpose. But with this approach, we will not know runtime upgrades as we continue to use the same cached one. I feel the entire `RuntimeCode` struct fields require heap allocation one way or the other considering the runtime upgrades.

I have a different approach in mind and would need a rebase. Moving this draft until then.

nazar-pc reviewed (commented):
> No we removed the dependency of system domain client, created a stateless runtime, invoked a necessary apis. Though I'm not sure I follow you.
>
> Initally, I was creating the Executor instance and instantiating same WASM everytime. Changed that to use the exisiting executor instead.

I mean you could just have something as simple as `Fn(&Extrinsic) -> Option<StateRoots>` rather than requiring `MessengerApi`. Just a simple abstract function/trait method that can do what you need to be done. How it is implemented wouldn't matter too much. Could be backed by runtime, could be a simple function.

vedhavyas committed:
define state root extractor and provide an impl with system domain client.

This is the main crust of the PR. Beacuse we want to use XDM verifier during the fraud proof verification on primary chain client but node may or may not be running the system domain which is currently used to extract state roots using System domain runtime. We provide a trait bound to extract the state roots so that we can inject system domain client or other implementation from primary chain.

vedhavyas committed:
provides an implementation of the StateRootExtractor with just system domain runtime WASM.

This implementation can be used on primary chain while doing fraud proof verification. The runtime code is expected as argument since we would have extracted the runtime from primary chain for fraud proof anyway. As for the executor, we can reuse the exisiting one on the primary chain.

Note: this implementation will be used in next PRs.

vedhavyas committed:
move verification module and state root extraction to sp_messenger and rename generic to better match its purpose

vedhavyas reviewed (commented):
@nazar-pc @liuchengxu I have updated the PR description. Start from there to understand the changes

nazar-pc commented on domains/client/domain-executor/src/state_root_extractor.rs:
```diff
@@ -0,0 +1,60 @@
+            .and_then(|maybe_state_roots| {
+                maybe_state_roots.ok_or(ApiError::Application("Empty state roots".into()))
+            })
+    }
```
This assumes best hash can decode the extrinsic, but is it the case though? What if we had runtime upgrade and latest runtime can no longer decode older extrinsic because it is gone now. I feel like there should be some context provided into this method such that decoding can be done properly.

vedhavyas replied:
I have updated the api to expect a context if any other implementation may need it.
But this is not useful with our current requirements and implementations. 
- We use best hash when we have access to system domain client because the context of usage is verifying extrinsincs in TXpool and Bundleexecution where we want to use the best_hash.
- For stateless wasm invokation, we dont need the context as long as runtime provided was capable since that is mainly used to setup the necessary backend.

nazar-pc reviewed (commented):
Is this PR complete? I see `StateRootExtractorWithSystemDomainRuntime` is introduced, but not used anywhere.

It seems like the same heap allocation would be necessary with this new approach, so I'm not sure what has fundamentally changed here. Also caching doesn't mean we blindly cache the first runtime, we certainly can follow runtime upgrades and make sure to not return outdated information from cache.

vedhavyas committed:
update StateRootExtractor to pass block hash from the caller.

When we use syatem domain client to extract state roots when verifying TXN using TX pool and Bundleverifier which are assumed to use best hash of the chain. When we use system domain runtime, we expect caller to provide a valid hash which is used to decode the extrinsic.

Block hash is only used to setup the right backend and is not used for anything when runtime is used.

nazar-pc reviewed (approved):
So if I understand correctly we still need to allocate memory for the extracted runtime, hash it, but only when we get a fraud proof, which, supposedly, isn't very often. If that is correct then it looks fine to me.

vedhavyas committed:
Merge branch 'main' into messeneger/state_root_extraction
